backend/src/apiserver/server/visualization_server.go

```diff
@@ -0,0 +1,111 @@
+package server
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"github.com/kubeflow/pipelines/backend/api/go_client"
+	"github.com/kubeflow/pipelines/backend/src/apiserver/resource"
+	"github.com/kubeflow/pipelines/backend/src/common/util"
+	"io/ioutil"
+	"net/http"
+	"net/url"
+	"strings"
+)
+
+type VisualizationServer struct {
+	resourceManager *resource.ResourceManager
+	serviceURL      string
+}
+
+func (s *VisualizationServer) CreateVisualization(ctx context.Context, request *go_client.CreateVisualizationRequest) (*go_client.Visualization, error) {
+	if err := s.validateCreateVisualizationRequest(request); err != nil {
+		return nil, err
+	}
+	body, err := s.GenerateVisualizationFromRequest(request)
+	if err != nil {
+		return nil, err
+	}
+	request.Visualization.Html = string(body)
+	return request.Visualization, nil
+}
+
+// validateCreateVisualizationRequest ensures that a go_client.Visualization
+// object has valid values.
+// It returns an error if a go_client.Visualization object does not have valid
+// values.
+func (s *VisualizationServer) validateCreateVisualizationRequest(request *go_client.CreateVisualizationRequest) error {
+	if len(request.Visualization.InputPath) == 0 {
+		return util.NewInvalidInputError("A visualization requires an InputPath to be provided. Received %s", request.Visualization.InputPath)
+	}
+	// Manually set Arguments to empty JSON if nothing is provided. This is done
+	// because visualizations such as TFDV and TFMA only require an InputPath to
+	// provided for a visualization to be generated. If no JSON is provided
+	// json.Valid will fail without this check as an empty string is provided for
+	// those visualizations.
+	if len(request.Visualization.Arguments) == 0 {
+		request.Visualization.Arguments = "{}"
+	}
+	if !json.Valid([]byte(request.Visualization.Arguments)) {
+		return util.NewInvalidInputError("A visualization requires valid JSON to be provided as Arguments. Received %s", request.Visualization.Arguments)
+	}
+	return nil
+}
+
+// getArgumentsAsJSONFromRequest will convert the values within a
+// go_client.CreateVisualizationRequest object to valid JSON that can be used to
+// pass arguments to the python visualization service.
+// It returns the generated JSON as an array of bytes and any error that is
+// encountered.
+func (s *VisualizationServer) getArgumentsAsJSONFromRequest(request *go_client.CreateVisualizationRequest) ([]byte, error) {
```

Comment:
is this function still needed? :)

Reply:
Yes, it is used within `createPythonArgumentsFromRequest` to create a string with all the arguments required for the python service.